src/Database/PostgreSQL/PQTypes/SQL/Builder.hs

```diff
@@ -408,9 +415,28 @@ instance Sqlable SqlInsertSelect where
     , emitClausesSepComma "RETURNING" $ sqlInsertSelectResult cmd
     ]
 
+checkAndRememberMaterializationSupport :: (MonadDB m, MonadIO m, MonadMask m) => m ()
+checkAndRememberMaterializationSupport = do
+  res :: Either DBException Int64 <- try . withNewConnection $ do
+    runSQL01_ $ "WITH t(n) AS MATERIALIZED (SELECT 1) SELECT n FROM t LIMIT 1"
+    fetchOne runIdentity
+  liftIO $ writeIORef withMaterializedSupported (isRight res)
```

Comment:
I wonder if it wasn't better to just use CPP here based on the postgres library instead of using IORef. Moving this isssue of "is it supported?" to compile time seems to be a better solution to me simply because it's a problem of a dependent library version.

Reply:
I don't think it's about library version - it doesn't seem to me that libpq does any syntax checking before sending query to the server. Still the idea here is to be able to interact with any version with underlying postgresql server without recompilation.

Reply:
> the idea here is to be able to interact with any version with underlying postgresql server without recompilation.

Exactly, we need to be able to switch from PostgreSQL 11 to 14 without recompilation and (hopefully_ without altering query plans due to a change in how WITH clauses are treated. Having said that, it would be good to have this as a comment in the code :)

Reply:
Yep, never mind... I wasn't thinking too much (or maybe was thinking way too much) when I wrote that comment.